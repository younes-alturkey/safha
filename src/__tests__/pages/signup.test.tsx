import '@testing-library/jest-dom'
import { cleanup, fireEvent, render, waitFor } from '@testing-library/react'
import { act } from 'react-dom/test-utils'
import TestWrapper from 'src/@core/components/test-wrapper'
import SignUpPage from 'src/pages/signup'

const pushMock = jest.fn()

jest.mock('next/router', () => ({
  useRouter: () => ({
    push: pushMock,
    events: {
      on: jest.fn(),
      off: jest.fn(),
      emit: jest.fn()
    },
    isFallback: false
  })
}))

jest.mock('next/navigation', () => ({
  useSearchParams: () => ({
    get: jest.fn()
  })
}))

jest.mock('src/api/auth', () => ({
  register: jest.fn().mockResolvedValue({ status: 201 })
}))

jest.mock('next-auth/react', () => ({
  signIn: jest.fn().mockResolvedValue({ status: 200 }),
  useSession: jest.fn().mockReturnValue({
    data: {
      user: {
        email: 'admin@safha.com'
      }
    }
  })
}))

jest.mock('src/api/analytics', () => ({
  createEvent: jest.fn().mockResolvedValue({ status: 200, data: {} }),
  getIPInfo: jest.fn().mockResolvedValue({ status: 200, data: {} })
}))

let container: HTMLElement
let firstnameInput: HTMLInputElement
let lastnameInput: HTMLInputElement
let emailInput: HTMLInputElement
let passwordInput: HTMLInputElement
let submitButton: HTMLButtonElement
let togglePasswordButton: HTMLButtonElement
let logoImage: HTMLImageElement
const random = Math.random().toString(36).substring(2, 15)
const user = {
  firstname: 'Test',
  lastname: 'User',
  email: `${random}@safha.com`,
  password: 'Safha123..'
}

const redirectParams = '/dash'

beforeEach(() => {
  jest.useFakeTimers()

  const rendered = render(
    <TestWrapper mode='light'>
      <SignUpPage apiUrl='https://sapi.safha.com' />
    </TestWrapper>
  )
  container = rendered.container
  firstnameInput = document.getElementById('firstname') as HTMLInputElement
  lastnameInput = document.getElementById('lastname') as HTMLInputElement
  emailInput = document.getElementById('email') as HTMLInputElement
  passwordInput = document.getElementById('password') as HTMLInputElement
  submitButton = document.getElementById('signup-button') as HTMLButtonElement
  togglePasswordButton = document.getElementById('toggle-password') as HTMLButtonElement
  logoImage = document.getElementById('safha-logo') as HTMLImageElement
})

afterEach(() => {
  jest.runOnlyPendingTimers()
  jest.useRealTimers()
  cleanup()
})

describe('SignUp Page', () => {
  it('renders SignUp page unchanged', () => {
    expect(container).toMatchSnapshot()
  })

  it('renders submit button correctly', () => {
    expect(submitButton).toBeInTheDocument()
  })

  it('renders submit button disabled', () => {
    expect(submitButton).toBeDisabled()
  })

  it('displays the logo image', () => {
    expect(logoImage).toBeInTheDocument()
  })

  it('allows user to input email and password', async () => {
    await act(async () => {
      fireEvent.change(firstnameInput, { target: { value: user.firstname } })
      fireEvent.change(lastnameInput, { target: { value: user.lastname } })
      fireEvent.change(emailInput, { target: { value: user.email } })
      fireEvent.change(passwordInput, { target: { value: user.password } })
    })

    expect(firstnameInput).toHaveValue(user.firstname)
    expect(lastnameInput).toHaveValue(user.lastname)
    expect(emailInput).toHaveValue(user.email)
    expect(passwordInput).toHaveValue(user.password)
  })

  it('registers a user successfully', async () => {
    await act(async () => {
      fireEvent.change(firstnameInput, { target: { value: user.firstname } })
      fireEvent.change(lastnameInput, { target: { value: user.lastname } })
      fireEvent.change(emailInput, { target: { value: user.email } })
      fireEvent.change(passwordInput, { target: { value: user.password } })
      fireEvent.click(submitButton)
    })

    await waitFor(() => expect(pushMock).toHaveBeenCalledWith(redirectParams))
  })

  it('toggles password visibility', async () => {
    fireEvent.click(togglePasswordButton)
    expect(passwordInput).toHaveAttribute('type', 'text')

    fireEvent.click(togglePasswordButton)
    expect(passwordInput).toHaveAttribute('type', 'password')
  })
})
