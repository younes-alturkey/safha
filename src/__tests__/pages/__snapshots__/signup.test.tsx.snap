// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUp Page renders SignUp page unchanged 1`] = `
<div>
  <div
    class="layout-wrapper MuiBox-root css-33gw4"
  >
    <div
      class="app-content MuiBox-root css-10qwnza"
    >
      <div
        class="content-center MuiBox-root css-imtqqp"
      >
        <div
          class="MuiBox-root css-1i2bp0u"
        >
          <div
            class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation7 MuiCard-root css-tt7mc4-MuiPaper-root-MuiCard-root"
          >
            <div
              class="MuiCardContent-root css-10pa0r4-MuiCardContent-root"
            >
              <div
                class="MuiBox-root css-fdeg8e"
              >
                <img
                  alt="Safha Logo"
                  data-nimg="1"
                  data-testid="safha-logo"
                  decoding="async"
                  height="75"
                  id="safha-logo"
                  loading="lazy"
                  src="/_next/image?url=https%3A%2F%2Fstorage.googleapis.com%2Fsafha-public-01%2Fsafha-logo-transparent-black.png&w=384&q=75"
                  srcset="/_next/image?url=https%3A%2F%2Fstorage.googleapis.com%2Fsafha-public-01%2Fsafha-logo-transparent-black.png&w=256&q=75 1x, /_next/image?url=https%3A%2F%2Fstorage.googleapis.com%2Fsafha-public-01%2Fsafha-logo-transparent-black.png&w=384&q=75 2x"
                  style="color: transparent; opacity: 0.5;"
                  width="150"
                />
              </div>
              <form
                action="#"
                autocomplete="off"
                class="w-full"
                novalidate=""
              >
                <div
                  class="MuiBox-root css-10vvd19"
                >
                  <div
                    class="MuiBox-root css-79elbk"
                  >
                    <div
                      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-11ns3xm-MuiFormControl-root-MuiTextField-root"
                      data-testid="firstname"
                    >
                      <label
                        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-filled MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-filled css-7peqn8-MuiFormLabel-root-MuiInputLabel-root"
                        data-shrink="true"
                        for="firstname"
                        id="firstname-label"
                      >
                        الاسم الأول
                      </label>
                      <div
                        class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-ah3eer-MuiInputBase-root-MuiFilledInput-root"
                      >
                        <input
                          aria-invalid="false"
                          class="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputSizeSmall css-8wyvbo-MuiInputBase-input-MuiFilledInput-input"
                          id="firstname"
                          placeholder="اسمك الأول"
                          type="text"
                          value=""
                        />
                      </div>
                    </div>
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-1msog3x-MuiTypography-root"
                    />
                  </div>
                  <div
                    class="MuiBox-root css-79elbk"
                  >
                    <div
                      class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-11ns3xm-MuiFormControl-root-MuiTextField-root"
                      data-testid="lastname"
                    >
                      <label
                        class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-filled MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-filled css-7peqn8-MuiFormLabel-root-MuiInputLabel-root"
                        data-shrink="true"
                        for="lastname"
                        id="lastname-label"
                      >
                        الاسم الأخير
                      </label>
                      <div
                        class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-ah3eer-MuiInputBase-root-MuiFilledInput-root"
                      >
                        <input
                          aria-invalid="false"
                          class="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputSizeSmall css-8wyvbo-MuiInputBase-input-MuiFilledInput-input"
                          id="lastname"
                          placeholder="اسمك الأخير"
                          type="text"
                          value=""
                        />
                      </div>
                    </div>
                    <p
                      class="MuiTypography-root MuiTypography-body1 css-1msog3x-MuiTypography-root"
                    />
                  </div>
                </div>
                <div
                  class="MuiBox-root css-79elbk"
                >
                  <div
                    class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-11ns3xm-MuiFormControl-root-MuiTextField-root"
                    data-testid="email"
                  >
                    <label
                      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-filled MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-filled css-7peqn8-MuiFormLabel-root-MuiInputLabel-root"
                      data-shrink="true"
                      for="email"
                      id="email-label"
                    >
                      البريد الإلكتروني
                    </label>
                    <div
                      class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall css-ah3eer-MuiInputBase-root-MuiFilledInput-root"
                    >
                      <input
                        aria-invalid="false"
                        class="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputSizeSmall css-8wyvbo-MuiInputBase-input-MuiFilledInput-input"
                        id="email"
                        placeholder="younes@safha.com"
                        type="text"
                        value=""
                      />
                    </div>
                  </div>
                  <p
                    class="MuiTypography-root MuiTypography-body1 css-1msog3x-MuiTypography-root"
                  />
                </div>
                <div
                  class="MuiBox-root css-79elbk"
                >
                  <div
                    class="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-1hsdtsd-MuiFormControl-root-MuiTextField-root"
                    data-testid="password"
                  >
                    <label
                      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-filled MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-filled css-7peqn8-MuiFormLabel-root-MuiInputLabel-root"
                      data-shrink="true"
                      for="password"
                      id="password-label"
                    >
                      كلمة المرور
                    </label>
                    <div
                      class="MuiInputBase-root MuiFilledInput-root MuiFilledInput-underline MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd css-xw93yy-MuiInputBase-root-MuiFilledInput-root"
                    >
                      <input
                        aria-invalid="false"
                        class="MuiInputBase-input MuiFilledInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd css-13scjpl-MuiInputBase-input-MuiFilledInput-input"
                        id="password"
                        placeholder="············"
                        type="password"
                        value=""
                      />
                      <div
                        class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-filled MuiInputAdornment-sizeSmall css-tsj3tr-MuiInputAdornment-root"
                      >
                        <button
                          aria-label="toggle password visibility"
                          class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-xubj68-MuiButtonBase-root-MuiIconButton-root"
                          data-testid="toggle-password"
                          id="toggle-password"
                          tabindex="0"
                          type="button"
                        >
                          <span />
                          <span
                            class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                          />
                        </button>
                      </div>
                    </div>
                  </div>
                  <p
                    class="MuiTypography-root MuiTypography-body1 css-1msog3x-MuiTypography-root"
                  />
                </div>
                <button
                  class="MuiButtonBase-root MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth Mui-disabled MuiButton-root MuiLoadingButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-fullWidth css-i5lb15-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
                  data-testid="signup-button"
                  disabled=""
                  id="signup-button"
                  tabindex="-1"
                  type="submit"
                >
                  إنشاء حساب
                </button>
                <div
                  class="MuiBox-root css-z0xj7h"
                >
                  <a
                    class="MuiTypography-root MuiTypography-body1 css-1my6stc-MuiTypography-root"
                    href="/signin"
                  >
                    تسجيل الدخول
                  </a>
                </div>
              </form>
            </div>
            <button
              aria-label="switch app locale"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-hub4ho-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <span />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              aria-label="switch app locale"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-hmdvb8-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <span />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
